cmake/configs/modes.cmake

```diff
@@ -9,104 +9,173 @@
 # Establishes build configuration modes we can use when compiling.
 
 include(CheckCCompilerFlag)
+include(CheckCXXCompilerFlag)
+include(${CMAKE_SOURCE_DIR}/cmake/helpers.cmake)
 
-set(build_modes None Debug Release)
-if("${CMAKE_C_COMPILER_ID}" STREQUAL "MSVC")
-    set(no_omit_frame_flag "/Oy-")
-else()
-    set(no_omit_frame_flag "-fno-omit-frame-pointer")
-endif()
+# Establish an internal cache variable to track our custom build modes.
+set(BUILD_MODES None Debug Release CACHE INTERNAL "")
```

Comment:
I am always confused about when we should use the `FORCE` flag and when we should not, could you explain why it is not necessary here?

Reply:
Yep, fair question. The `FORCE` flag is used when you want to override the value of the cache variable and ensure it stays that value in the CMake cache i.e if `set(x "foo" FORCE)` and `set(x "bar" FORCE)` were executed in sequence, variable `x` would stay as "foo" (unless another `FORCE` was used). Its useful to ensure a cache variable value doesn't get modified or you want to override a user set cache value (perhaps due to some configuration state)
We don't use it here, since `BUILD_MODES` will be dynamically appended to and modified

Reply:
I think in your example, you meant `set(x "bar")`.
It makes sense, thanks!

Reply:
Yep, sorry, I made a typo.